gr-digital/lib/costas_loop_cc_impl.cc

```diff
@@ -193,6 +206,72 @@ int costas_loop_cc_impl::work(int noutput_items,
         phase_wrap();
         frequency_limit();
 
+        if (hasAdditionalOutputs) {
+            if (freq_optr != NULL)
+                freq_optr[i] = d_freq;
+            if (phase_optr != NULL)
+                phase_optr[i] = d_phase;
+            if (error_optr != NULL)
+                error_optr[i] = d_error;
+        }
+    }
+
+    return noutput_items;
+}
+
+int costas_loop_cc_impl::work2(int noutput_items,
+                               gr_vector_const_void_star& input_items,
+                               gr_vector_void_star& output_items)
+{
+    const gr_complex* iptr = (gr_complex*)input_items[0];
+    gr_complex* optr = (gr_complex*)output_items[0];
+    float* freq_optr = output_items.size() >= 2 ? (float*)output_items[1] : NULL;
+    float* phase_optr = output_items.size() >= 3 ? (float*)output_items[2] : NULL;
+    float* error_optr = output_items.size() >= 4 ? (float*)output_items[3] : NULL;
+
+    float n_r, n_i, o_r, o_i;
+
+    std::vector<tag_t> tags;
+    get_tags_in_range(tags,
+                      0,
+                      nitems_read(0),
+                      nitems_read(0) + noutput_items,
+                      pmt::intern("phase_est"));
+
+    for (int i = 0; i < noutput_items; i++) {
+        if (!tags.empty()) {
+            if (tags[0].offset - nitems_read(0) == (size_t)i) {
+                d_phase = (float)pmt::to_double(tags[0].value);
+                tags.erase(tags.begin());
+            }
+        }
+
+        if ((d_phase > M_TWOPI) || (d_phase < CL_MINUS_TWO_PI)) {
+            d_phase =
+                d_phase * CL_ONE_OVER_2PI - (float)((int)(d_phase * CL_ONE_OVER_2PI));
+            d_phase = d_phase * M_TWOPI;
+        }
+        n_i = sinf(-d_phase);
+        n_r = cosf(-d_phase);
+
+        o_r = (iptr[i].real() * n_r) - (iptr[i].imag() * n_i);
+        o_i = (iptr[i].real() * n_i) + (iptr[i].imag() * n_r);
+
+        optr[i].real(o_r);
+        optr[i].imag(o_i);
+
+        d_error = o_r * o_i;
+        d_error = 0.5 * (std::abs(d_error + 1) - std::abs(d_error - 1));
```

Comment:
This looks like ``branchless_clip()``. No need to inline manually, the compiler does. People reading the code will wonder what this line makes (at least to me it is not obvious).